File: flowersec-swift/Sources/Flowersec/TransportSecurity.swift
import Foundation

#if canImport(Darwin)
  import Darwin
#elseif canImport(Glibc)
  import Glibc
#endif

public enum TransportRuntime: String, Codable, Equatable, Sendable {
  case swift
}

public struct TransportSecurityPolicyInput: Equatable, Sendable {
  public var path: FlowersecPath
  public var scheme: String
  public var host: String
  public var runtime: TransportRuntime

  public init(path: FlowersecPath, scheme: String, host: String, runtime: TransportRuntime) {
    self.path = path
    self.scheme = scheme
    self.host = host
    self.runtime = runtime
  }
}

public enum PlaintextRiskAcceptance: String, Equatable, Sendable {
  case acceptPreE2ECredentialExposure = "accept_pre_e2ee_credential_exposure"
}

public struct NetworkPlaintextPolicyOptions: Equatable, Sendable {
  public var allowedHosts: [String]
  public var riskAcceptance: PlaintextRiskAcceptance

  public init(allowedHosts: [String], riskAcceptance: PlaintextRiskAcceptance) {
    self.allowedHosts = allowedHosts
    self.riskAcceptance = riskAcceptance
  }
}

public enum TransportSecurityPolicy: Sendable {
  case requireTLS
  case allowPlaintextForLoopback
  @available(
    *, deprecated,
    message: "Use requireTLS, allowPlaintextForLoopback, or networkPlaintext(options:)."
  )
  case allowPlaintext
  case custom(@Sendable (TransportSecurityPolicyInput) async throws -> Bool)

  public static func networkPlaintext(options: NetworkPlaintextPolicyOptions) throws -> Self {
    guard options.riskAcceptance == .acceptPreE2ECredentialExposure else {
      throw NetworkPlaintextPolicyError.invalidRiskAcceptance
    }
    guard !options.allowedHosts.isEmpty else {
      throw NetworkPlaintextPolicyError.missingAllowedHosts
    }
    let hosts = try Set(options.allowedHosts.map(canonicalNetworkPlaintextHost))
    return .custom { input in
      input.scheme == "wss" || (input.scheme == "ws" && hosts.contains(input.host))
    }
  }
}

private enum NetworkPlaintextPolicyError: Error {
  case invalidRiskAcceptance
  case missingAllowedHosts
  case invalidAllowedHost(String)
}

private func stringFromNullTerminatedBuffer(_ buffer: [CChar]) -> String {
  String(
    decoding: buffer.prefix { $0 != 0 }.map { UInt8(bitPattern: $0) },
    as: UTF8.self
  )
}

private func canonicalNetworkPlaintextHost(_ rawHost: String) throws -> String {
  let host = rawHost.trimmingCharacters(in: .whitespacesAndNewlines)
  guard !host.isEmpty,
    host == host.lowercased(),
    host.rangeOfCharacter(from: CharacterSet(charactersIn: "@/?#%[]")) == nil
  else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
  }

  if host.contains(":") {
    var address = in6_addr()
    guard inet_pton(AF_INET6, host, &address) == 1 else {
      throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
    }
    var buffer = [CChar](repeating: 0, count: Int(INET6_ADDRSTRLEN))
    let rendered = buffer.withUnsafeMutableBufferPointer { output in
      withUnsafePointer(to: &address) { pointer in
        inet_ntop(AF_INET6, UnsafeRawPointer(pointer), output.baseAddress, socklen_t(output.count))
      }
    }
    guard rendered != nil else {
      throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
    }
    let canonical = stringFromNullTerminatedBuffer(buffer)
    guard canonical == host else {
      throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
    }
    let words = try expandIPv6Words(canonical)
    let unspecified = words.allSatisfy { $0 == 0 }
    let loopback = words.dropLast().allSatisfy { $0 == 0 } && words.last == 1
    let mappedIPv4 = words.prefix(5).allSatisfy { $0 == 0 } && words[5] == 0xffff
    let multicast = words[0] & 0xff00 == 0xff00
    let linkLocal = words[0] & 0xffc0 == 0xfe80
    guard !unspecified, !loopback, !mappedIPv4, !multicast, !linkLocal else {
      throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
    }
    return canonical
  }

  var address = in_addr()
  guard inet_pton(AF_INET, host, &address) == 1 else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
  }
  var buffer = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
  let rendered = buffer.withUnsafeMutableBufferPointer { output in
    withUnsafePointer(to: &address) { pointer in
      inet_ntop(AF_INET, UnsafeRawPointer(pointer), output.baseAddress, socklen_t(output.count))
    }
  }
  guard rendered != nil else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
  }
  let canonical = stringFromNullTerminatedBuffer(buffer)
  guard canonical == host else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
  }
  let octets = host.split(separator: ".").compactMap { Int($0) }
  guard octets.count == 4,
    octets[0] != 127,
    !(octets[0] == 169 && octets[1] == 254),
    octets[0] < 224,
    !octets.allSatisfy({ $0 == 0 })
  else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(rawHost)
  }
  return canonical
}

private func expandIPv6Words(_ host: String) throws -> [UInt16] {
  let halves = host.components(separatedBy: "::")
  guard halves.count <= 2 else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(host)
  }
  let left = halves[0].isEmpty ? [] : halves[0].split(separator: ":").map(String.init)
  let right =
    halves.count == 1 || halves[1].isEmpty ? [] : halves[1].split(separator: ":").map(String.init)
  let missing = 8 - left.count - right.count
  guard (halves.count == 1 && missing == 0) || (halves.count == 2 && missing > 0) else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(host)
  }
  let parts = left + Array(repeating: "0", count: missing) + right
  let words = parts.compactMap { UInt16($0, radix: 16) }
  guard words.count == 8 else {
    throw NetworkPlaintextPolicyError.invalidAllowedHost(host)
  }
  return words
}

public struct TransportSecurityDiagnostic: Equatable, Sendable {
  public var code: String
  public var path: FlowersecPath
  public var scheme: String
  public var host: String
  public var runtime: TransportRuntime

  public init(
    code: String,
    path: FlowersecPath,
    scheme: String,
    host: String,
    runtime: TransportRuntime
  ) {
    self.code = code
    self.path = path
    self.scheme = scheme
    self.host = host
    self.runtime = runtime
  }
}

enum FlowersecTransportSecurity {
  static func enforce(url: URL, path: FlowersecPath, options: ConnectOptions) async throws {
    let target: (scheme: String, host: String)
    do {
      target = try parse(url)
    } catch {
      throw denied(path: path)
    }
    let input = TransportSecurityPolicyInput(
      path: path,
      scheme: target.scheme,
      host: target.host,
      runtime: .swift
    )

    let allowed: Bool
    do {
      switch options.transportSecurityPolicy {
      case .requireTLS:
        allowed = target.scheme == "wss"
      case .allowPlaintextForLoopback:
        allowed = target.scheme == "wss" || isLiteralLoopbackHost(target.host)
      case .allowPlaintext:
        allowed = true
      case .custom(let evaluate):
        try Task.checkCancellation()
        allowed = try await evaluate(input)
        try Task.checkCancellation()
      }
    } catch is CancellationError {
      throw CancellationError()
    } catch {
      if Task.isCancelled { throw CancellationError() }
      throw denied(path: path)
    }
    if !allowed {
      throw denied(path: path)
    }
    if target.scheme == "ws" {
      options.onDiagnosticEvent?(
        DiagnosticEvent(
          path: path,
          stage: .transport,
          codeDomain: .event,
          code: "plaintext_transport",
          result: .skip,
          resource: "websocket_transport"
        )
      )
      options.onTransportSecurityDiagnostic?(
        TransportSecurityDiagnostic(
          code: "plaintext_transport",
          path: path,
          scheme: target.scheme,
          host: target.host,
          runtime: .swift
        )
      )
    }
  }

  private static func denied(path: FlowersecPath) -> FlowersecError {
    FlowersecError(
      path: path,
      stage: .validate,
      code: .transportPolicyDenied,
      message: "Transport security policy denied the WebSocket URL."
    )
  }

  private static func parse(_ url: URL) throws -> (scheme: String, host: String) {
    let raw = url.absoluteString.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let separator = raw.range(of: "://") else { throw ParseError.invalid }
    let scheme = raw[..<separator.lowerBound].lowercased()
    guard scheme == "ws" || scheme == "wss" else { throw ParseError.invalid }
    let suffix = raw[separator.upperBound...]
    let authority = String(suffix.prefix { $0 != "/" && $0 != "?" && $0 != "#" })
    guard !authority.isEmpty, !authority.contains("@") else { throw ParseError.invalid }

    let host: String
    if authority.hasPrefix("[") {
      guard let end = authority.firstIndex(of: "]"), end > authority.startIndex else {
        throw ParseError.invalid
      }
      host = String(authority[authority.index(after: authority.startIndex)..<end]).lowercased()
      let remainder = authority[authority.index(after: end)...]
      if !remainder.isEmpty {
        guard remainder.first == ":", remainder.dropFirst().allSatisfy(\.isNumber) else {
          throw ParseError.invalid
        }
      }
    } else {
      let pieces = authority.split(separator: ":", omittingEmptySubsequences: false)
      guard pieces.count <= 2 else { throw ParseError.invalid }
      if pieces.count == 2 {
        guard !pieces[1].isEmpty, pieces[1].allSatisfy(\.isNumber) else {
          throw ParseError.invalid
        }
      }
      host = String(pieces[0]).lowercased()
    }
    guard !host.isEmpty else { throw ParseError.invalid }
    return (scheme, host)
  }

  private static func isLiteralLoopbackHost(_ host: String) -> Bool {
    if host == "localhost" || host == "::1" { return true }
    let pieces = host.split(separator: ".", omittingEmptySubsequences: false)
    guard pieces.count == 4 else { return false }
    var octets: [Int] = []
    for piece in pieces {
      let value = String(piece)
      guard !value.isEmpty,
        value == "0" || !value.hasPrefix("0"),
        value.allSatisfy(\.isNumber),
        let octet = Int(value),
        octet <= 255
      else { return false }
      octets.append(octet)
    }
    return octets.first == 127
  }

  private enum ParseError: Error {
    case invalid
  }
}

File: flowersec-ts/src/client-connect/transportSecurity.ts
import type { ClientPath } from "../client.js";
import { FlowersecError } from "../utils/errors.js";
import { emitObserverDiagnostic, type ClientObserverLike } from "../observability/observer.js";

export const RequireTLS = "require_tls" as const;
export const AllowPlaintextForLoopback = "allow_plaintext_for_loopback" as const;
/** @deprecated Use RequireTLS, AllowPlaintextForLoopback, or createNetworkPlaintextPolicy. */
export const AllowPlaintext = "allow_plaintext" as const;

export const PlaintextRiskAcceptance = {
  acceptPreE2ECredentialExposure: "accept_pre_e2ee_credential_exposure",
} as const;

export type PlaintextRiskAcceptance = typeof PlaintextRiskAcceptance[keyof typeof PlaintextRiskAcceptance];

export type NetworkPlaintextPolicyOptions = Readonly<{
  allowedHosts: readonly string[];
  riskAcceptance: PlaintextRiskAcceptance;
}>;

export type TransportSecurityPolicyInput = Readonly<{
  path: ClientPath;
  scheme: "ws" | "wss";
  host: string;
  runtime: "browser" | "node" | "other";
}>;

export type TransportSecurityPolicyPreset =
  | typeof RequireTLS
  | typeof AllowPlaintextForLoopback
  | typeof AllowPlaintext;

export type TransportSecurityPolicy =
  | TransportSecurityPolicyPreset
  | ((input: TransportSecurityPolicyInput) => boolean | Promise<boolean>);

export function createNetworkPlaintextPolicy(options: NetworkPlaintextPolicyOptions): TransportSecurityPolicy {
  if (options?.riskAcceptance !== PlaintextRiskAcceptance.acceptPreE2ECredentialExposure) {
    throw new Error("network plaintext policy requires explicit pre-E2EE credential exposure acceptance");
  }
  if (!Array.isArray(options.allowedHosts) || options.allowedHosts.length === 0) {
    throw new Error("network plaintext policy requires at least one allowed host");
  }
  const allowedHosts = new Set(options.allowedHosts.map(canonicalNetworkPlaintextHost));
  return (input) => input.scheme === "wss" || (input.scheme === "ws" && allowedHosts.has(input.host));
}

type ParsedWebSocketTarget = Readonly<{
  scheme: "ws" | "wss";
  host: string;
}>;

export async function enforceTransportSecurity(args: Readonly<{
  rawUrl: string;
  path: ClientPath;
  policy?: TransportSecurityPolicy;
  observer?: ClientObserverLike;
}>): Promise<void> {
  let target: ParsedWebSocketTarget;
  try {
    target = parseWebSocketTarget(args.rawUrl);
  } catch (cause) {
    throw denied(args.path, cause);
  }
  const input: TransportSecurityPolicyInput = {
    path: args.path,
    scheme: target.scheme,
    host: target.host,
    runtime: detectRuntime(),
  };

  let allowed = false;
  const policy = args.policy ?? RequireTLS;
  try {
    if (typeof policy === "function") {
      allowed = await policy(input);
    } else {
      allowed = evaluatePreset(policy, target);
    }
  } catch (cause) {
    throw denied(args.path, cause);
  }
  if (!allowed) throw denied(args.path);
  if (target.scheme === "ws") {
    emitObserverDiagnostic(args.observer, {
      path: args.path,
      stage: "transport",
      code_domain: "event",
      code: "plaintext_transport",
      result: "skip",
    });
  }
}

function denied(path: ClientPath, cause?: unknown): FlowersecError {
  return new FlowersecError({
    path,
    stage: "validate",
    code: "transport_policy_denied",
    message: "transport security policy denied websocket URL",
    ...(cause === undefined ? {} : { cause }),
  });
}

function evaluatePreset(policy: TransportSecurityPolicyPreset, target: ParsedWebSocketTarget): boolean {
  switch (policy) {
    case RequireTLS:
      return target.scheme === "wss";
    case AllowPlaintextForLoopback:
      return target.scheme === "wss" || isLiteralLoopbackHost(target.host);
    case AllowPlaintext:
      return true;
  }
}

function parseWebSocketTarget(rawUrl: string): ParsedWebSocketTarget {
  const raw = rawUrl.trim();
  const match = /^([A-Za-z][A-Za-z0-9+.-]*):\/\/([^/?#]*)(?:[/?#]|$)/.exec(raw);
  const scheme = match?.[1]?.toLowerCase();
  const authority = match?.[2] ?? "";
  if ((scheme !== "ws" && scheme !== "wss") || authority === "" || authority.includes("@")) {
    throw new Error("invalid websocket URL");
  }

  let host: string;
  if (authority.startsWith("[")) {
    const end = authority.indexOf("]");
    if (end <= 1 || (authority.slice(end + 1) !== "" && !/^:\d+$/.test(authority.slice(end + 1)))) {
      throw new Error("invalid websocket URL");
    }
    host = authority.slice(1, end).toLowerCase();
  } else {
    const pieces = authority.split(":");
    if (pieces.length > 2 || (pieces.length === 2 && !/^\d+$/.test(pieces[1] ?? ""))) {
      throw new Error("invalid websocket URL");
    }
    host = (pieces[0] ?? "").toLowerCase();
  }
  if (host === "") throw new Error("invalid websocket URL");
  return { scheme, host };
}

function isLiteralLoopbackHost(host: string): boolean {
  if (host === "localhost" || host === "::1") return true;
  const parts = host.split(".");
  if (parts.length !== 4) return false;
  const octets: number[] = [];
  for (const part of parts) {
    if (!/^(0|[1-9]\d{0,2})$/.test(part)) return false;
    const value = Number(part);
    if (value > 255) return false;
    octets.push(value);
  }
  return octets[0] === 127;
}

function canonicalNetworkPlaintextHost(rawHost: string): string {
  const host = String(rawHost ?? "").trim();
  if (host === "" || host !== host.toLowerCase() || /[@/?#%\[\]]/.test(host)) {
    throw new Error(`invalid network plaintext allowed host ${JSON.stringify(rawHost)}`);
  }
  if (host.includes(":")) {
    return canonicalNetworkIPv6Host(host);
  }
  const octets = canonicalIPv4Octets(host);
  const firstOctet = octets[0];
  const secondOctet = octets[1];
  if (firstOctet == null || secondOctet == null) {
    throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(rawHost)}`);
  }
  if (firstOctet === 127 || (firstOctet === 169 && secondOctet === 254) || firstOctet >= 224 || octets.every((value) => value === 0)) {
    throw new Error(`network plaintext allowed host must be a non-loopback unicast IP literal: ${JSON.stringify(rawHost)}`);
  }
  return host;
}

function canonicalIPv4Octets(host: string): readonly number[] {
  const parts = host.split(".");
  if (parts.length !== 4) {
    throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(host)}`);
  }
  const octets = parts.map((part) => {
    if (!/^(0|[1-9]\d{0,2})$/.test(part)) {
      throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(host)}`);
    }
    const value = Number(part);
    if (value > 255) {
      throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(host)}`);
    }
    return value;
  });
  return octets;
}

function canonicalNetworkIPv6Host(host: string): string {
  let parsed: URL;
  try {
    parsed = new URL(`http://[${host}]/`);
  } catch {
    throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(host)}`);
  }
  const canonical = parsed.hostname.replace(/^\[|\]$/g, "").toLowerCase();
  if (canonical !== host) {
    throw new Error(`network plaintext allowed host must be a canonical IP literal: ${JSON.stringify(host)}`);
  }
  const words = expandIPv6Words(canonical);
  const unspecified = words.every((word) => word === 0);
  const loopback = words.slice(0, 7).every((word) => word === 0) && words[7] === 1;
  const mappedIPv4 = words.slice(0, 5).every((word) => word === 0) && words[5] === 0xffff;
  const multicast = (words[0]! & 0xff00) === 0xff00;
  const linkLocal = (words[0]! & 0xffc0) === 0xfe80;
  if (unspecified || loopback || mappedIPv4 || multicast || linkLocal) {
    throw new Error(`network plaintext allowed host must be a non-loopback unicast IP literal: ${JSON.stringify(host)}`);
  }
  return canonical;
}

function expandIPv6Words(host: string): readonly number[] {
  const halves = host.split("::");
  if (halves.length > 2) throw new Error("invalid IPv6 host");
  const left = halves[0] === "" ? [] : halves[0]!.split(":");
  const right = halves.length === 1 || halves[1] === "" ? [] : halves[1]!.split(":");
  const missing = 8 - left.length - right.length;
  if ((halves.length === 1 && missing !== 0) || (halves.length === 2 && missing < 1)) {
    throw new Error("invalid IPv6 host");
  }
  const words = [...left, ...Array.from({ length: missing }, () => "0"), ...right].map((part) => Number.parseInt(part, 16));
  if (words.length !== 8 || words.some((word) => !Number.isInteger(word) || word < 0 || word > 0xffff)) {
    throw new Error("invalid IPv6 host");
  }
  return words;
}

function detectRuntime(): TransportSecurityPolicyInput["runtime"] {
  if (typeof window !== "undefined" && typeof window.document !== "undefined") return "browser";
  if (typeof process !== "undefined" && process.versions?.node != null) return "node";
  return "other";
}
